gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/FeatureConverter.kt

```diff
@@ -0,0 +1,144 @@
+            fillFeature(builder, project, doc.id, f)
+            return builder.build()
+        }
+
+        f.location?.let {
+            val builder = PointFeature.newBuilder().setPoint(toPoint(it))
+            fillFeature(builder, project, doc.id, f)
+            return builder.build()
+        }
+
+        throw DataStoreException("No geometry in remote feature " + doc.id)
+    }
+
+    private fun hasNonEmptyVertices(featureDocument: FeatureDocument): Boolean {
+        val geometry = featureDocument.geometry
+
+        if (geometry == null
+            || geometry[GEOMETRY_COORDINATES] == null
+            || geometry[GEOMETRY_COORDINATES] !is List<*>
+        ) {
+            return false
+        }
+
+        val coordinates = geometry[GEOMETRY_COORDINATES] as List<*>?
+        return coordinates?.isNotEmpty() ?: false
+    }
+
+    private fun toFeatureFromGeometry(
+        project: Project, doc: DocumentSnapshot, f: FeatureDocument
+    ): PolygonFeature {
+        val geometry = f.geometry
+        val type = geometry!![GEOMETRY_TYPE]
+        if (POLYGON_TYPE != type) {
+            throw DataStoreException("Unknown geometry type in feature " + doc.id + ": " + type)
+        }
+
+        val coordinates = geometry[GEOMETRY_COORDINATES]
+        if (coordinates !is List<*>) {
+            throw DataStoreException(
+                "Invalid coordinates in feature " + doc.id + ": " + coordinates
+            )
+        }
+
+        val vertices = ImmutableList.builder<Point>()
+        for (point in coordinates) {
+            if (point !is GeoPoint) {
+                Timber.d("Ignoring illegal point type in feature %s", doc.id)
+                break
+            }
+            vertices.add(
+                Point.newBuilder().setLongitude(point.longitude).setLatitude(
+                    point.latitude
+                ).build()
+            )
+        }
+
+        val builder = PolygonFeature.builder().setVertices(vertices.build())
+        fillFeature(builder, project, doc.id, f)
+        return builder.build()
+    }
+
+    private fun fillFeature(
+        builder: Feature.Builder<*>, project: Project, id: String, f: FeatureDocument
+    ) {
+        val layerId = DataStoreException.checkNotNull(f.layerId, LAYER_ID)
+        val layer =
+            DataStoreException.checkNotEmpty(project.getLayer(layerId), "layer " + f.layerId)
+        // Degrade gracefully when audit info missing in remote db.
+        val created = f.created ?: AuditInfoNestedObject.FALLBACK_VALUE
+        val lastModified = f.lastModified ?: created
+        builder
+            .setId(id)
+            .setProject(project)
+            .setCustomId(f.customId)
+            .setCaption(f.caption)
+            .setLayer(layer)
+            .setCreated(AuditInfoConverter.toAuditInfo(created))
+            .setLastModified(AuditInfoConverter.toAuditInfo(lastModified))
+    }
+
+    private fun toPoint(geoPoint: GeoPoint): Point? =
```

Comment:
Should this be non-nullable?

Reply:
I don't think so, it looks like it was annotated as @Nullable before, but it doesn't seem necessary